Add new course for R reproducible environments (#26)
Public-Health-Scotland/learnr-online: ReproducibleEnvs.Rmd

-   Be able to confidently apply reproducibility principles to improve reliability and transparency of projects.

### What is `{renv}`?

R manages packages through a central library and shares this across all projects. In PHS Posit Workbench, we expand on this slightly, with a system-level library (used for base packages) and then user-level where each user-installed package will go. On any system, this location can be found using the `.libPaths()` function, where the list of locations is listed in order of priority.

This works well until versions of packages are considered. Each project may require different versions of packages, resulting in deprecated or missing functionality. This issue is only made worse the moment the project is shared with someone else, someone with their very own library of packages with different versions. The `{renv}` package solves this issue by creating a separate, reproducible environment that is attached to the project, creating:

-   a project-specific library, and
-   a list of dependencies and their version.

```{r renv-libraries, echo=FALSE, fig.align='center'}
knitr::include_graphics("images/renv-libraries.png")

> ```

Reviewer: I found this graphic a bit confusing, I think probably some form of nesting would convey the concept better... 

I had a quick go at a nested hex (fun little diversion), see code and images below. If you agree with my point feel free to use the images/code or trash them and make better ones!

<details>
  <summary>Code to produce</summary>

```r

library(ggplot2)
library(phsstyles)
# Data frame for a hexagon ----
angle <- seq(0, 2 * pi, length.out = 7) + pi / 6
polygon_df <- data.frame(
  angle = angle,
  x     = cos(angle),
  y     = sin(angle)
)

# Colours ----
hex_col_1 <- phs_colors("phs-purple")
hex_col_2 <- phs_colors("phs-magenta")
hex_col_3 <- phs_colors("phs-blue")

transparent_white <- "#FFFFFF00"

# Plots with nested hexagons, sequentially dashed.
p_1 <- ggplot(polygon_df) +
  geom_polygon(
    aes(x = x, y = y),
    linetype = "dashed",
    fill = transparent_white,
    colour = hex_col_1
  ) +
  geom_polygon(
    aes(x = x * 2 / 3, y = y * 2 / 3),
    fill = transparent_white,
    colour = hex_col_2
  ) +
  geom_polygon(
    aes(x = x * 1 / 3, y = y * 1 / 3),
    fill = transparent_white,
    colour = hex_col_3
  ) +
  coord_equal() +
  theme_void()

p_2 <- ggplot(polygon_df) +
  geom_polygon(
    aes(x = x, y = y),
    fill = transparent_white,
    colour = hex_col_1
  ) +
  geom_polygon(
    aes(x = x * 2 / 3, y = y * 2 / 3),
    linetype = "dashed",
    fill = transparent_white,
    colour = hex_col_2
  ) +
  geom_polygon(
    aes(x = x * 1 / 3, y = y * 1 / 3),
    fill = transparent_white,
    colour = hex_col_3
  ) +
  coord_equal() +
  theme_void()

p_3 <- ggplot(polygon_df) +
  geom_polygon(
    aes(x = x, y = y),
    fill = transparent_white,
    colour = hex_col_1
  ) +
  geom_polygon(
    aes(x = x * 2 / 3, y = y * 2 / 3),
    fill = transparent_white,
    colour = hex_col_2
  ) +
  geom_polygon(
    aes(x = x * 1 / 3, y = y * 1 / 3),
    linetype = "dashed",
    fill = transparent_white,
    colour = hex_col_3
  ) +
  coord_equal() +
  theme_void()

p_1
p_2
p_3

# Save plots ----
ggsave("outer_hex.png", p_1)
ggsave("middle_hex.png", p_2)
ggsave("inner_hex.png", p_3)
```
</details>

![inner_hex](https://github.com/Public-Health-Scotland/learnr-online/assets/5982260/423bffbd-c6e5-42af-834d-a409cb8dcf33)
![middle_hex](https://github.com/Public-Health-Scotland/learnr-online/assets/5982260/cdf36de1-f2b9-4c8b-a79b-72b5073a2f6a)
![outer_hex](https://github.com/Public-Health-Scotland/learnr-online/assets/5982260/804b7834-df20-4ce0-bf11-34a91bf9b357)

Author: Hey, @Moohan - thanks for your review, I've implemented changes based on your suggestions. 
The current graphic isn't great, I created it to have something visual but I don't like it either so I'm keen to explore alternatives. While I do like the style of your suggestions, I'm hesitant to show the model as a nested one. I understand where you're coming from but I'm concerned it's misleading from how libraries actually work. I'll have a think...

</br>

#### Projects

The next major component of setting up a project for reproducibility is using RStudio Projects. R accesses a working directory, found using `getwd()`, and can be changed using `setwd()` (not recommended). This is likely to change between users but also between your projects which could cause confusion and issue when switching from one bit of work to another. This can be set up as below from File > New Project (we'll come back to setting up projects with renv - the green box):

```{r renv-newproj, echo=FALSE, fig.align='center'}
knitr::include_graphics("images/renv-newproj1.png")
```

Building on the solid foundations that RStudio Projects provide, we need to use relative paths within our code. A relative path is *relative* to the current working directory (set by the project). This means that if an output is saved using a relative path: `data/project_data.csv`, different users would access the same location regardless of where their working version of the project (i.e. `project`) is stored, e.g.:

* `/Projects/user1/project/data/project_data.csv`
* `/Projects/user2/phs_projects/project/data/project_data.csv`

> While less of an issue in PHS as we all use Posit Workbench to access RStudio, there can also be differences introduced depending on operating systems. If you're working with external collaborators for instance, they may use a Windows system which would require different formatting for file paths. 

Reviewer: Possibly worth including here, that even within PHS Posit WB we can have different versions of R - 3 at the moment and these are likely (guaranteed?) to change / increase over time.

```{r renv-proj, echo=FALSE, fig.align='center'}
knitr::include_graphics("images/illustrations/r_cracked_setwd.png")
```

### Environments

Understanding environments in this context is the next fundamental. In the context of using computers to handle and analyse data, to developing outputs such as dashboards and applications, an environment refers to the collection of software tools, libraries, configurations, and dependencies that are required to run a particular program or project. Environments can vary based on factors such as programming language, project requirements, and external dependencies. Also, while we use Posit Workbench in PHS, this eliminates some problems as we're all starting from the same point, but each user can then set their own configuration and packages that will inevitably be different from the next person. Managing these environments effectively is crucial for ensuring that workflows are reproducible across time but also between users and, ultimately, across systems.

```{r renv-env, echo=FALSE, fig.align='center', out.width='100%'}
knitr::include_graphics("images/renv-positenv.png")
```

> Using Posit Workbench means that our local environment has little impact on any development work. The main difference that can occur between projects and users for R projects is packages, which packages and which version of those packages. That's where `{renv}` comes in, so lets get a look at how to set it up...

Reviewer: ```suggestion
Using Posit Workbench means that our local environment has little impact on any development work. The main difference that can occur between projects and users for R projects is packages, which packages and which version of those packages. That's where `{renv}` comes in, so let's look at how to set it up.
```

## Setting up `{renv}`

> The `{renv}` package is a package like any other, so this needs to be installed in order to work with it. This package is available through the PHS Package Manager (or generally through CRAN) and can be installed using `install.packages("renv")`. On PHS Posit Workbench, this will be installed in your user library as standard, this is then used from here to generate the project level library.

Reviewer: ```suggestion
`{renv}` is a package like any other, so must first be installed. It is available through the PHS Package Manager (or generally through CRAN) and can be installed using `install.packages("renv")`. On PHS Posit Workbench, this will be installed in your user library as standard, this is then used to generate the project level library.
```
